### DISCUSSION TITLE
vips conversion to jp2k is changing colors. how to preserve colors?

### DISCUSSION THREAD

**Discussion by jrochkind** [libvips/libvips#3428] [Q&A]
OK, now that I have vips succesfully creating jp2 files, I'm playing around with it, and I see some oddities with color in documents converted from TIFF to jp2. vips 8.14.2

We start with this TIFF: https://jrochkind.github.io/vips_jp2_color_example/dhc6a4r.1200.tiff

Let's first make an ordinary JPEG conversion. We run:  `vips copy dhc6a4r.1200.tiff dhc6a4r.vips.jpeg` It results in this: https://jrochkind.github.io/vips_jp2_color_example/dhc6a4r.vips.jpeg

If we compare that JPEG to the TIFF, the colors look identical to me. 

If we instead convert to a jpeg200:  `vips copy dhc6a4r.1200.tiff dhc6a4r.vips.jp2`, that results in this: https://jrochkind.github.io/vips_jp2_color_example/dhc6a4r.vips.jp2

If we look at the jp2 side-by-side with the TIFF or the jpeg -- the colors on the jp2 seem a little bit different, they are a little bit blue-er, they have changed. 

Does anyone know what's going on here, or if there's a way to convert a TIFF to jp2 and keep the colorspace/colors the same? 

if i try the same thing with imagemagick: `dhc6a4r.1200.tiff dhc6a4r.vips.jp2` -- that results in this: https://jrochkind.github.io/vips_jp2_color_example/dhc6a4r.imck.jp2 . Which also has the very same color change, it looks identical to the vips-produced jp2, a little bit blue-er than the original TIFF or JPEG. 

So the issue is not unique to vips.   I'm not really sure what's going on, but would like to figure out a way to convert to jp2 without the color shfit. Is there a way to tell the jp2 to use the same colorspace (? not sure if this is even the right thing?) and colors as the TIFF?

**Comment by jcupitt** [comment]
Hi again, your image has an embedded ICC profile:

```
$ vipsheader -a dhc6a4r.1200.tiff | grep icc
icc-profile-data: 560 bytes of binary data
```

jpeg2000 (or at least openjpeg) doesn't support embedded profiles, so it's being lost. 

If you try:

```
vips copy dhc6a4r.1200.tiff dhc6a4r.vips.jpeg[strip]
```

Then convert that to jp2 you should find they match.

**Reply by jrochkind** [reply]
Hi, sorry to come back to this three months later, I got pulled to something else -- but appreciate your answer a lot and am back!  Thank you for pointing out it's a color profile issue -- that makes sense. 

Digital colorspace/profiles are definitely confusing to me, but I did some reading to try to get at least a little of a refreshed handle on it. 

It looks like JPEG2000 itself may actually support embeded ICC color profiles. It looks like while originally [only certain kinds of profiles were supported](https://wiki.opf-labs.org/display/TR/Handling+of+ICC+profiles) (I'm not certain how limiting this ends up being, or if the profile in for example above image would qualify), with some [significant expansion](https://openpreservation.org/blogs/icc-profiles-and-resolution-jp2-update-2011-d-lib-paper/) in 2013. 

But if OpenJPEG doesn't support any of this, and only commercial JP2 libraries (which I don't use and dont' want to) do, okay, there we are. 

My understanding of ICC profiles is that they basically are specifications of "color spaces". Such as sRGB is a color space (that doesn't need a profile, since it's a standard). 

I am not sure if there is a _conversion_ going on when I am converting from TIFF with a profile to jp2 (without profile) -- or if it's just stripping color profile without doing any color conversion?  Is there any color conversion that would be appropriate to try to convert from the color profile to... whatever color space is standard in jp2? (sRGB?).   Or is this even misunderstanding what the possiblities are?

Basically, I'm wondering, is there anything that would be possible to try to convert and preserve the colors as much as possible -- or is this already what's going on, and the changes I am seeing are as much as possible?

I would not expect such drastic visible changes from dropping the color profile, if the proper converesion was done -- but my expectations may be wrong!

I was trying to experiment with converting the TIFF to sRGB _without_ embedded color profile, before then converting that to jp2 -- to see if that made any difference -- but I'm not sure if I succesfully figured out the right commands to make that happen, or if that's even a reasonable thing to do. 

Any guidance or advice appreciated!  And thank you for your work on vips which is just so amazing, and your work to support people and answer questions, which I notice and appreciate!

**Reply by jcupitt** [reply]
Yes, the convention is generally "no profile == srgb", so you could try:

```
$ vips icc_transform input.jpg output.jp2 srgb
```

Now that'll apply the embedded profile to the JPG pixels, then write those srgb pixel values to the jp2 file and not attach a profile.

**Reply by jcupitt** [reply]
I tried with your image:

```
$ vips copy ../dhc6a4r.1200.tiff x.jpg
$ vips icc_transform ../dhc6a4r.1200.tiff x.jp2 srgb
$ vips copy x.jp2 x2.jpg
```

So `x.jpg` is your TIFF with an embedded profile but as a JPG with an embedded profile:

![x](https://github.com/libvips/libvips/assets/580843/06244bd0-9be8-41a4-af9b-c5cb27b1f5c1)

And `x2.jpg` is your image encoded as srgb jp2 with no profile, then copied back to a JPG with no profile:

![x2](https://github.com/libvips/libvips/assets/580843/bdb6f658-101f-4408-b72c-cb556bedb045)

If you put those two images in a pair of brower tabs and flip between them, they should look identical (though the numeric pixel values are different).

**Reply by jrochkind** [reply]
Oh awesome, this is great!

I spent some hours yesterday trying to figure this out, and hadn't been able to get there yet. 

I am currently home sick with a bad cold or something, but next week I will try this out too!

Thank you so much for your help.

**Reply by jrochkind** [reply]
OK, I have confirmed this, indeed, as you say.  Excellent!

I somehow had not managed to figure out how to do this actual color space conversion (actual change of bytes) on my own -- with vips CLI (hadn't discovered the `iccprofile` command), with imagemagick, even with Photoshop GUI, I hadn't managed to make it happen. But your recipe seems to do it. 

I further tried to use [jpylyzer](https://jpylyzer.openpreservation.org/) to confirm what was going on in the "Colour Specification Box" header metadata in the jp2s -- there ARE multiple colour spaces supported by .jp2, even before we get to ICC profiles. They all look like this, which seems consistent iwth them being set to sRGB, I hope...

```
            <colourSpecificationBox>
                <meth>Enumerated</meth>
                <prec>0</prec>
                <approx>0</approx>
                <enumCS>sRGB</enumCS>
            </colourSpecificationBox>
```


## Mystery of what is going with those images, and profiles

One mystery to me -- if I take the former mis-colored JP2, and open it up in MacOS Preview, and choose `Tools` / `Assign Profile` ("Assign ColorSync Profile to Image") / `sRGB IEC61966-2.1` -- it seems to work to correct the colors. I believe this is only assigning ICC profile metadata, and not actually converting bytes... if it wasn't already sRGB, how does this possibly work?  (Perhaps there are different sRGB profiles involved? But that doesn't make sense to me either).  Remains a mystery. 

## Can this be done in one command?

OK, I can resize (to fixed width, while keeping aspect ratio the same) and convert between formats in one command possibly a couple different ways with vips CLI, including:  `vipsthumbnail   orig.tiff --size 1400x65500  -o output.jpg` 

To resize, change format (to jp2), AND convert to sRGB... is there any way to do this in _one_ command? I am going to do this in bulk, so avoiding hitting the disk for the extra output/input could be non-trivial...

The [vipsthumbnail documentation](https://www.libvips.org/API/current/Using-vipsthumbnail.html) has a "colour management" section... that suggests to me `--export-profile srgb` maybe should do it (how did I not find that before?)

Seems to have worked! GREAT. How did I not find that before?  

## Should vips have warnings/more options?

If vips knows you are going to be losing color info, as I did in my original naive command, should it output a warning?

But I guess maybe it is hard to keep track of when this will be the case, as it kind of depends on the matrix of what different readers/writers are going to do/support? Fair enough. 

If I read the docs right, `export-profile` converts pixel bytes, and ALSO strips ICC profile... is/should there be a way to convert pixel bytes to a new profile AND ALSO embed that new profile as ICC profile?

Should `vips copy` have an option that lets you convert color profiles like this too? Support the `--export-profile` option, etc?

## On thumbnail generation

Once we started paying attention to this unintended color shfit, we realized our default thumbnail generation was subject to it too (it looked like almost the SAME color shift as the initial faulty jp2 generation)

We had based our recipe for thumbnail generation (a long time ago) on http://libvips.blogspot.com/2013/11/tips-and-tricks-for-vipsthumbnail.html , and _thought_ it was converting to sRGB as well as stripping the profile, with `--eprofile /path/to/sRGB.icc  --delete`

But apparently it never was, we were mistaken all along, I think this was faulty advice? Or is `--export-profile` and `--eprofile` synonyms, and it _should_ have worked all along? Pretty confused. 

And I see now, as I said the docs at https://www.libvips.org/API/current/Using-vipsthumbnail.html suggest using `--export-profile srgb` with the built-in profile (rather than `--eprofile` with a bring-your-own ICC srgb), and without `--delete` but with `-o tn_shark.jpg[profile=none]`.... 

I will update to new advice and see if it fixes my problems there....


## OK, but dzsave....

We also use `vips dzsave` to create DZI tiles, I'm wondering if dzsave might do anything under the hood that might not to be up to date with properly preserving/converting colour space info?

This and the last one both definitely need more investigation on my end; feel free to ignore for now, I'll come back with a separate item at some point after investigating more, if I haven't figured it out.

**Comment by jrochkind** [comment]
@jcupitt 

So.... trying `vipsthumbnail orig.tiff --resize 900x65500 ---export-profile srgb -o out.jp2`....

On my MacOS machine with vips `8.14.2` -- this works fine. 

On another machine that is at the moment stuck on vips `8.12.1` because ubuntu... for the exact same input,  it is _for some but not all input images_ producing weird greyscale output (that wasn't originally greyscale) in the produced jp2 thumbs.  (and another machine with vips 8.13.0 also reproduces the problem). 

It oddly only effects _some_ images... and also only **some** sizes. If I resize the same image to 1200x65500 it does _not_ reproduce problem, if I resize to 900x65500, it does. 

Looking at the CHANGELOG, I _do_ see some things between `8.13.0` and `8.14.2` that could be relevant.  Such as especially `- fix ICC handling of RGB images with a monochrome profile [kleisauke]`. That could be my case. 

Do you have a gut intuition of whether I have run into a bug in older vips that has been fixed in newer?  (It occurs to me that there could also be differnet verisons of OpenJPEG going on, or who knows what).

**Reply by jcupitt** [reply]
From memory it's to do with chroma subsampling -- openjpeg does not do this well for odd image sizes. Recent libvipses have defaulted to having this disabled.

**Reply by jrochkind** [reply]
Awesome, thank you. So a more recent libvips probably WILL fix my problem.  This is very helpful thanks so much for info. 

Wow, working with JP2 (at least with OpenJPEG) has so many more gotchas than I expected, I may retreat from it.

**Reply by jcupitt** [reply]
It is badly unloved :( The commercial jp2k libraries (Kakadu etc.) are much better, but of course you have to pay a lot of $$$ for them. It's why jp2k is only really used in high value vertical markets, like medical imaging.

**Reply by jrochkind** [reply]
@jcupitt I'm in an environment (heroku) where installing the most recent libvips is somewhat challenging (but not impossible). 

But I had a thought... 

> From memory it's to do with chroma subsampling -- openjpeg does not do this well for odd image sizes. Recent libvipses have defaulted to having this disabled.

OK, recent verisons have _defaulted_....

if I'm using an older version that may not default (say `8.13.0` or `8.12.1`) -- is there any way on the commandline for `vipsthumbnail` that I can specify to disable chroma subsampling for jp2 save?

Sorry, I feel bad keeping you here, but appreciate your help so much! I am trying to look in vips docs for ways to disable chroma subsampling on jp2 save from `vipsthumbnail`, but am not quite sure where to look, although maybe i'm getting closer...

**Reply by jcupitt** [reply]
Sure, `subsample-mode=off`, I think.

**Reply by jrochkind** [reply]
HECK yeah, that seems to have gotten me back in business, working around the problem. 

    vipsthumbnail orig.tiff --size 900x65500 --export-profile srgb output.jp2[Q=40,subsample-mode=off]

Thank you so much jcupitt!  Your over-and-above support to users is noticed and appreciated, you are an open source legend. 

FWIW for the record, I found the reported issue you mentioned, yup: https://github.com/libvips/libvips/issues/2965
